### ISSUE TITLE
Const generic equivalent of `Curve::batch_normalize`

### ISSUE THREAD

**Issue by tarcieri** [zkcrypto/group#50]
`batch_normalize` currently documents this caveat:

> This function will panic if `p.len() != q.len()`.

However, with const generics you can construct an output array that matches the input size with compile-time assurances:

```rust
fn batch_normalize_array<const N: usize>(points: &[Self, N]) -> [Self::AffineRepr; N] { ... }
```

**Comment by tarcieri** [comment]
We ended up solving this with a [`BatchNormalize` trait](https://docs.rs/elliptic-curve/0.13.7/elliptic_curve/point/trait.BatchNormalize.html) which is generic around its input type and therefore supports overlapping impls.

This allows the trait to be impl'd for `[ProjectivePoint]` and `[ProjectivePoint; N]` as inputs, where the former is gated on the `alloc` feature, and the latter supporting heapless usage, with the const generics completely disentangled from the trait definition: https://docs.rs/primeorder/0.13.6/primeorder/struct.ProjectivePoint.html#impl-BatchNormalize%3C%5BProjectivePoint%3CC%3E%5D%3E-for-ProjectivePoint%3CC%3E
